IPNI.md

```diff
@@ -463,6 +463,11 @@ and uses the multihash portion of the CID only.
 * `cid` - _Required_. The default string representation of the Cid. Currently, Base32 is used for
   CIDv1 as the encoding for the multibase string and Base58 is used for CIDv0.
 
+##### Query Parameters
+
+* `cascade` - _Optional_. The comma separated alternative routing systems to which the lookup is cascaded in addition to searching IPNI index records. If unspecified, only IPNI index records are searched. Supported values:
```

Comment:
Allowing for multiple parameters without the user being able to know which are available or were used to satisfy a request still leaves implementations a little clueless about what to do.

For example, if I don't know if cid.contact has an IPFS Public DHT plugin and I query it how will I know if I should also search the IPFS Public DHT?

It seems like in order for this to work that you'd want a "what do I speak" endpoint. You might also want to tag responses with the set of routers you checked.

Reply:
> Today, the ask is to optionally search the DHT without doubling traffic on the DHT when requests come from Kubo. I think the proposed changes here deliver on that.

Technically this is correct even if cascading was mandatory since Kubo doesn't use this API, but instead the `/routing/v1` API (the fact that the indexstar implementation of `/routing/v1` uses IPNI APIs to backend implementations internally is somewhat coincidental. However, I get your point that this does the minimal thing.

---

What I suspect you'd want to enable here is that developers (or users) could choose to forgo public DHT requests entirely and just send them all to cid.contact if they wanted, and only do the requests themselves if cid.contact was failing/unavailable for some reason. In order for them to do this in the current proposal, they'd need to hard-code "hey, cid.contact speaks dht" somewhere in their code/config.

This is the same kind of thing that is likely to cause user problems in https://github.com/ipfs/specs/pull/342.